IPC::Connection message throttling refers to main run loop

kkinnunen-apple opened the pull request:
#### 50386bbd03a77eb8e157d4cab3c9d8a1cb10794a
<pre>
IPC::Connection message throttling refers to main run loop
<a href="https://bugs.webkit.org/show_bug.cgi?id=246276">https://bugs.webkit.org/show_bug.cgi?id=246276</a>
rdar://problem/100974660

Reviewed by Chris Dumez.

Using main run loop in the MessageThrottler makes it hard to
make Connection run loop agnostic.

Instead, just remove MessageThrottler, it represents just one
int. Store the state to Connection.

Use RunLoop::dispatch() for both cases: when the messages are
throttled and when they are not. Zero interval timers and
RunLoop::dispatchAfter(0_s) are dispatched within same
run loop iteration, and thus the test cannot test that
the throttling works as expected. The purpose of the zero
interval timer was supposedly give time for running
other tasks in run loop, but it might actually make
the dispatch task earlier than the already submitted
non-timer tasks.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::enableIncomingMessagesThrottling):
(IPC::Connection::getPendingMessageCountForTesting):
(IPC::Connection::numberOfMessagesToProcess):
(IPC::Connection::dispatchIncomingMessages):
(IPC::Connection::MessagesThrottler::MessagesThrottler): Deleted.
(IPC::Connection::MessagesThrottler::scheduleMessagesDispatch): Deleted.
(IPC::Connection::MessagesThrottler::numberOfMessagesToProcess): Deleted.
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::isIncomingMessagesThrottlingEnabled const):
(IPC::Connection::isThrottlingIncomingMessages const):
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
(TestWebKitAPI::TEST_P):

Canonical link: <a href="https://commits.webkit.org/255579@main">https://commits.webkit.org/255579@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/98f968b7b6d270bbe39855082aaecb53fad79706

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92997 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2199 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23571 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102710 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162949 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96999 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2209 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30527 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85382 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98824 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98663 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/1512 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79472 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28429 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83426 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71541 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36924 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/17049 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34735 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18247 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3866 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38608 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40843 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40533 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37435 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -355,14 +355,16 @@ class Connection : public ThreadSafeRefCounted<Connection, WTF::DestructionThrea
     // Can be called from any thread.
     void dispatchDidReceiveInvalidMessage(MessageName);
+
+    size_t getPendingMessageCountForTesting();
```
Per WebKit coding style, we don't use the "get" prefix for getters in WebKit. This should be named `pendingMessageCountForTesting()`.

cdumez replied:
Also, can this be const? Seems odd for a getter to not be const. We may have to mark the Lock as mutable.

cdumez commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -448,11 +438,11 @@ class Connection : public ThreadSafeRefCounted<Connection, WTF::DestructionThrea
     bool m_fullySynchronousModeIsAllowedForTesting { false };
     bool m_ignoreTimeoutsForTesting { false };
     bool m_didReceiveInvalidMessage { false };
+    int m_incomingMessagesThrottlingLevel { -1 };
```
I think we're trying to move away from magic values in WebKit. This may look nicer as an std::optional<unsigned>.

cdumez commented on Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
```diff
@@ -55,6 +55,14 @@ class MockConnectionClient final : public IPC::Connection::Client {
 
+    Vector<MessageInfo> takeMessages()
+    {
+        Vector<MessageInfo> result;
```
This is testing code so I don't care much but it seems we could avoid memory allocations like so:
```
auto result = std::exchange(m_messages, Vector<MessageInfo>());
result.reverse();
return result;
```

kkinnunen-apple replied:
Used deque

cdumez commented on Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
```diff
@@ -77,10 +85,19 @@ class MockConnectionClient final : public IPC::Connection::Client {
+        MessageInfo messageInfo { decoder.messageName(), decoder.destinationID() };
+        if (m_asyncMessageHandler && m_asyncMessageHandler(messageInfo))
+            return;
+        m_messages.insert(0, WTFMove(messageInfo));
```
Again, this is testing code so probably doesn't matter but it seems like a good use case for a WTF::Deque, not a WTF::Vector when you have to prepend.

cdumez reviewed (approved):
LGTM

kkinnunen-apple committed:
IPC::Connection message throttling refers to main run loop
https://bugs.webkit.org/show_bug.cgi?id=246276
rdar://problem/100974660

Reviewed by Chris Dumez.

Using main run loop in the MessageThrottler makes it hard to
make Connection run loop agnostic.

Instead, just remove MessageThrottler, it represents just one
int. Store the state to Connection.

Use RunLoop::dispatch() for both cases: when the messages are
throttled and when they are not. Zero interval timers and
RunLoop::dispatchAfter(0_s) are dispatched within same
run loop iteration, and thus the test cannot test that
the throttling works as expected. The purpose of the zero
interval timer was supposedly give time for running
other tasks in run loop, but it might actually make
the dispatch task earlier than the already submitted
non-timer tasks.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::enableIncomingMessagesThrottling):
(IPC::Connection::getPendingMessageCountForTesting):
(IPC::Connection::numberOfMessagesToProcess):
(IPC::Connection::dispatchIncomingMessages):
(IPC::Connection::MessagesThrottler::MessagesThrottler): Deleted.
(IPC::Connection::MessagesThrottler::scheduleMessagesDispatch): Deleted.
(IPC::Connection::MessagesThrottler::numberOfMessagesToProcess): Deleted.
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::isIncomingMessagesThrottlingEnabled const):
(IPC::Connection::isThrottlingIncomingMessages const):
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
(TestWebKitAPI::TEST_P):

Canonical link: https://commits.webkit.org/255579@main
